pkg/pmem-csi-operator/metrics/metrics.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	// PmemCSIDeploymentSubsystemKey represents the key used for
+	// PMEM-CSI deployment metrics sub-system.
+	PmemCSIDeploymentSubsystemKey = "pmem_csi_deployment"
+)
+
+var (
+	// ReconcileStart creates new prometheus metrics counter
+	// that gets incremented on start of each reconcile loop
+	// of a PmemCSIDeployment CR, with information: {"name"}.
+	ReconcileStart = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: PmemCSIDeploymentSubsystemKey,
+		Name:      "reconcile_start",
+		Help:      "Number of reconcile loops gone through by a PmemCSIDeployment CR.",
+	}, []string{"name"})
+
+	// ReconcileEnd creates new prometheus metrics counter
+	// that gets incremented on end of each reconcile loop
+	// of a PmemCSIDeployment CR, with information: {"name"}.
+	ReconcileEnd = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: PmemCSIDeploymentSubsystemKey,
+		Name:      "reconcile_end",
+		Help:      "Number of reconcile loops gone through by a PmemCSIDeployment CR.",
+	}, []string{"name"})
+
+	// SubResourceCreatedAt creates new prometheus metrics for
+	// a sub resource deployed for a PmemCSIDeployment,
+	// with information: {"name", "namespace", "group", "version", "kind"}
+	SubResourceCreatedAt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Subsystem: PmemCSIDeploymentSubsystemKey,
+		Name:      "sub_resource_created_at",
+		Help:      "Timestamp at which a sub resource was created.",
+	}, []string{"name", "namespace", "group", "version", "kind"})
+
+	// SubResourceUpdatedAt creates new prometheus metrics for
+	// a sub resource redeployed for a PmemCSIDeployment,
+	// with information: {"name", "namespace", "group", "version", "kind"}
+	SubResourceUpdatedAt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Subsystem: PmemCSIDeploymentSubsystemKey,
+		Name:      "sub_resource_updated_at",
+		Help:      "Timestamp at which a sub resource was update.",
+	}, []string{"name", "namespace", "group", "version", "kind"})
+)
+
+func RegisterMetrics() {
+	metrics.Registry.MustRegister(
+		ReconcileStart,
+		ReconcileEnd,
+		SubResourceCreatedAt,
+		SubResourceUpdatedAt,
+	)
+}
+
+func SetSubResourceCreateMetric(obj client.Object) error {
+	if obj == nil {
+		return nil
+	}
+	return setGauge(SubResourceCreatedAt, GetSubResourceLabels(obj))
+}
+
+func SetSubResourceUpdateMetric(obj client.Object) error {
+	if obj == nil {
+		return nil
+	}
+	return setGauge(SubResourceUpdatedAt, GetSubResourceLabels(obj))
+}
+
+func DeleteSubResourceMetric(obj client.Object) {
```

Comment:
Why do we need to delete samples? 

Doesn't that introduce the risk that a metrics gatherer misses the information? Suppose a sub resource gets added and then deleted without any intermediate metrics gathering. The client will then not see the value for `created_at`.

Reply:
I suppose if the number of label tuples is unbounded, metrics data could grow without any bounds. But for our purposes I find that unlikely.